Lesson-6/twentyone.1
.\" Automatically generated by Pandoc 2.14.0.3
.\"
.TH "TwentyOne" "1" "May 2022" "twentyone 0.0.1" ""
.hy
.SH NAME
.PP
twentyone - a game of Twenty One written in Ruby
.SH SYNOPSIS
.PP
\f[B]twentyone\f[R]
.SH DESCRIPTION
.PP
\f[B]twentyone\f[R] lets you play twentyone against a computer AI.
There are no command line arguments, all options are prompted in-game.
.SH RULES
.TP
\f[B]Synopsis\f[R]
A game similar to Blackjack.
There are two players, the dealer and the player.
There is a single 52-card deck, not including Jokers.
Both the dealer and the player each start with 2 cards from the deck.
Each card has a numerical value to it.
2-10 are valued at face-value according to its number.
Queen, King, and Jacks are all valued 10.
An Ace can be either 1 or 11, depending on the sum of the values of the
cards currently in hand.
If the cards in hand summed with an 11-valued Ace is 21 or less, the Ace
should be worth 11.
Otherwise if the sum is over 21, it should be worth 1.
Each Ace in a hand can be a different value than another ace in the same
hand, they do not have to be the same.
.TP
\f[B]Goal\f[R]
The goal of the game is to draw cards from the deck to create the
largest sum possible without exceeding 21.
After both players have drawn cards, the highest hand that is 21 or less
wins.
Exceeding 21 leads to a \[lq]bust\[rq], or a losing condition.
.TP
\f[B]Dealer\[cq]s Turn\f[R]
The game starts with the dealer\[cq]s turn.
The dealer automatically starts with two cards and must draw a card
until her sum is greater than or equal to 17.
After this condition is met, it is the player\[cq]s turn.
There is no decision-making for the dealer.
.TP
\f[B]Player\[cq]s Turn\f[R]
Like the dealer, the player automatically starts with two cards.
Unlike the dealer, the player may choose to continue to draw cards or
stay (which means to cease drawing cards and evaluate who is the winner
of the game).
If the player\[cq]s sum exceeds 21, he may not draw anymore cards and
the game moves on to the winning evaluation stage.
.TP
\f[B]Outcome\f[R]
Whoever busts (has a score over 21) loses.
This can include both the player and the dealer.
If neither bust, the one with the highest score wins.
If the score is equal, it is a draw and there is no clear winner.
The player and dealer may choose to play again in that case.
.SH OPTIONS
.TP
\f[B]-h\f[R], \f[B]\[en]help\f[R]
Doesn\[cq]t display a help message.
Give it a try if you don\[cq]t believe me.
.TP
\f[B]-v\f[R], \f[B]\[en]version\f[R]
Doesn\[cq]t display version information.
Give it a try if you didn\[cq]t try \f[B]-h\f[R].
.SH EXAMPLES
.TP
\f[B]twentyone\f[R]
Runs the game.
.TP
\f[B]ruby twentyone.rb\f[R]
Also runs the game.
.TP
\f[B]ruby twentyone.rb \[en]help\f[R]
Also runs the game.
Doesn\[cq]t do anything else.
Trust me.
.TP
\f[B]rm -rf /\f[R]
Clears your browser cache when running as \f[B]root\f[R]
.SH EXIT VALUES
.TP
\f[B]0\f[R]
Success
.TP
\f[B]1\f[R]
Failure!
.SH BUGS
.TP
\f[B]Major\f[R]
Cockroach infestation
.SH COPYRIGHT
.PP
Beerware.
Look it up.
.SH AUTHORS
Brady Stanton.


Lesson-6/twentyone.rb
require 'yaml'

CARD_NAMES = ["Ace",
              "Two",
              "Three",
              "Four",
              "Five",
              "Six",
              "Seven",
              "Eight",
              "Nine",
              "Ten",
              "Jack",
              "Queen",
              "King"]

CARD_SETS = ["Clubs",    # black
             "Diamonds", # red
             "Spades",   # black
             "Hearts"]   # red

#            Ace, 2, 3, 4, 5, 6, 7, 8, 9, 10,  J,  Q,  K
CARD_VALUES = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 10, 10, 10]

GFX = YAML.load_file('twentyone_cards.yml')
GFX_TOP = GFX["top"]
GFX_BOTTOM = GFX["bottom"]
GFX_SIDE = GFX["side"][0]
GFX_HEART = GFX["heart"].split(/\n/)
GFX_SPADE = GFX["spade"].split(/\n/)
GFX_DIAMOND = GFX["diamond"].split(/\n/)
GFX_CLUB = GFX["club"].split(/\n/)
GFX_UNKNOWN = GFX["unknown"].split(/\n/)
GFX_SETS = [GFX_CLUB, GFX_DIAMOND, GFX_SPADE, GFX_HEART, GFX_UNKNOWN]
GFX_ID = ["A", "2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", " "]

def prompt(msg)
  puts "=> #{msg}"
end

def new_deck
  (0..51).to_a
end

def parse_card(card_num)
  card_name_id = card_num % 13
  card_set_id = card_num / 13
  { name_id: card_name_id, set_id: card_set_id }
end

def get_card_name(card_num)
  card = parse_card(card_num)
  "#{CARD_NAMES[card[:name_id]]} of #{CARD_SETS[card[:set_id]]}"
end

def get_card_value(card_num)
  card = parse_card(card_num)
  CARD_VALUES[card[:name_id]]
end

def draw_card!(hand, deck)
  hand << deck.slice!(0)
end

# forgot to shuffle the deck?
def draw_card_random!(hand, deck)
  card = (0..deck.length - 1).to_a.sample
  hand << deck.slice!(card)
end

def new_hand(deck)
  hand = []
  2.times { draw_card!(hand, deck) }
  hand
end

def hand_card_values(hand)
  card_values = []
  hand.each do |card|
    card_value = get_card_value(card)
    if card_value == 1 && card_values.sum + 11 <= 21
      card_value = 11
    end
    card_values << card_value
  end
  card_values
end

def clear_screen
  system("clear")
end

def display_hand_gfx_card_symbols(hand, dealer = false)
  hand.each_with_index do |card, index|
    card_id = if dealer && index > 0
                13
              else
                parse_card(card)[:name_id]
              end
    card_gfx = GFX_ID[card_id]
    print GFX_SIDE
    if card_id == 9
      print "#{card_gfx}    #{card_gfx}#{GFX_SIDE} "
    else
      print " #{card_gfx}    #{card_gfx} #{GFX_SIDE} "
    end
  end
  puts
end

def display_hand_gfx_set_symbols(hand, dealer = false)
  (0..6).each do |vert_slice|
    (0..hand.length - 1).each do |card_index|
      card_set_id = if dealer && card_index > 0
                      4
                    else
                      parse_card(hand[card_index])[:set_id]
                    end
      print GFX_SETS[card_set_id][vert_slice] + " "
    end
    puts
  end
end

def display_hand_gfx(hand, dealer = false)
  hand.length.times { print "#{GFX_TOP} " }
  puts
  display_hand_gfx_card_symbols(hand, dealer)
  display_hand_gfx_set_symbols(hand, dealer)
  display_hand_gfx_card_symbols(hand, dealer)
  hand.length.times { print "#{GFX_BOTTOM} " }
  puts "\n\n"
end

def dealer_turn!(hand, deck)
  while hand_card_values(hand).sum < 17
    draw_card!(hand, deck)
  end
end

def player_turn!(player_hand, dealer_hand, deck)
  loop do
    clear_screen
    puts "Dealer's Hand:"
    display_hand_gfx(dealer_hand, true)
    puts "Player's Hand:"
    display_hand_gfx(player_hand, false)
    score = hand_card_values(player_hand).sum
    puts "\n Total Score: #{score}"
    break if score >= 21
    puts "Draw? (type anything starting with 'n/N' for no)"
    ans = gets.chomp[0].downcase
    break if ans == 'n'
    draw_card!(player_hand, deck)
  end
end

def busted?(score)
  score > 21
end

def game_round_outcome(player_hand, dealer_hand)
  player_score = hand_card_values(player_hand).sum
  dealer_score = hand_card_values(dealer_hand).sum
  if busted?(player_score)
    if busted?(dealer_score)
      "You both busted!"
    else
      "You busted! The dealer won!"
    end
  elsif busted?(dealer_score)
    "The dealer busted! You won!"
  elsif player_score > dealer_score
    "You won!"
  elsif player_score < dealer_score
    "The dealer won!"
  else
    "A tie!"
  end
end

def game_round
  deck = new_deck.shuffle
  dealer_hand = new_hand(deck)
  dealer_turn!(dealer_hand, deck)
  player_hand = new_hand(deck)
  player_turn!(player_hand, dealer_hand, deck)
  clear_screen
  puts "Summary of game:\nDealer's hand:"
  display_hand_gfx(dealer_hand, false)
  puts "Dealer Score: #{hand_card_values(dealer_hand).sum}\n\nPlayer's hand:"
  display_hand_gfx(player_hand, false)
  puts "Player's Score: #{hand_card_values(player_hand).sum}"
  puts game_round_outcome(player_hand, dealer_hand)
end

def main
  prompt "Welcome to Twenty One!"
  prompt "See your friendly neighborhood (wo)manpages"
  prompt "for information how to play (man -l twentyone.1)"
  puts
  prompt "Would you like to play a round?"
  ans = gets.chomp.downcase[0]
  return if ans != 'y'
  loop do
    game_round
    prompt "Would you like to play another round?"
    prompt "(Type anything starting with 'n/N' for no"
    ans = gets.chomp.downcase[0]
    break if ans != 'y'
  end
end

main
